### ISSUE TITLE
Just checking on target mask...

### ISSUE THREAD

**Issue by dsantiago** [andrewpeng02/transformer-translation#7]
Hi there Andrew!

Just a question, isn't your mask wrong here:

https://github.com/andrewpeng02/transformer-translation/blob/d735f04bee6355d67c87f1ea3567de65f9291b46/train.py#L143-L151

i think it should be like 1's in place of 0's...

[1 -inf -inf]
[1     1 -inf]
[1     1     1]

**Comment by andrewpeng02** [comment]
Hi,

Where do you see that the 0's should be 1's? In the Pytorch documentation, "the non-zero positions will be ignored while the zero positions will be unchanged" [here](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html). The generate_square_subsequent_mask() function also uses 0's for unmasked and -inf for masked.

**Comment by dsantiago** [comment]
Maybe i misunderstood, and you aer setting 0 as False to mask just keep values...

2 implementations that i saw mask as True in the lower of the diagonal and False above

In this one:  http://nlp.seas.harvard.edu/2018/04/03/attention.html#background

Search for:
```python
def subsequent_mask(size):
    "Mask out subsequent positions."
    attn_shape = (1, size, size)
    subsequent_mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
    return torch.from_numpy(subsequent_mask) == 0
```

Or here also:

https://github.com/aladdinpersson/Machine-Learning-Collection/blob/1d78f390ab8ae803e6a56ef42cad506fc49b2c81/ML/Pytorch/more_advanced/transformer_from_scratch/transformer_from_scratch.py#L258-L262

**Comment by andrewpeng02** [comment]
I'm looking at the implementation of Pytorch's multihead attention here: [(note, v1.2)](https://github.com/pytorch/pytorch/blob/cb90e36684b638fd4cb33b1cdf23c22a96fd9806/torch/nn/functional.py#L3194-L3252). On line 3252, it adds ```attn_mask``` to ```attn_output_weights```, meaning that it the values with 0 don't change the corresponding value  ```attn_output_weights``` while the values with -inf essentially removes the target value in ```attn_output_weights``` and the model doesn't attend to that value. In v1.7, you can also use a BoolTensor and "positions with True is not allowed to attend while False values will be unchanged" (Pytorch docs), so 0 would correspond to False and -inf would correspond to True. 

I think that the issue is just a matter of implementation. For your second example at least, where the attention is defined [here](https://github.com/aladdinpersson/Machine-Learning-Collection/blob/1d78f390ab8ae803e6a56ef42cad506fc49b2c81/ML/Pytorch/more_advanced/transformer_from_scratch/transformer_from_scratch.py#L56-L57), the 0's (meaning the upper diagonal) are switched with -1e20 and the 1's are left unchanged. I'm not sure if leaving the 1's in the mask is intentional, but compared to -1e20, 1 is small enough to be basically 0 [EDIT: it doesn't add the 1, 1 just means the weights are left unchanged]. Does this make sense?

**Comment by dddvision** [comment]
I am seeing different versions of the masking.

A: Andrew's code here
https://github.com/andrewpeng02/transformer-translation/blob/d735f04bee6355d67c87f1ea3567de65f9291b46/train.py#L143-L151
produces FloatTensor:
[[  0. -inf -inf]
 [  0.   0. -inf]
 [  0.   0.   0.]]

B: Pytorch reference code for generate_square_subsequent_mask here
https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#Transformer.generate_square_subsequent_mask
produces FloatTansor:
[[  0. -inf -inf]
 [  0.   0. -inf]
 [  0.   0.   0.]]

C: The Harvard NLP code here
http://nlp.seas.harvard.edu/2018/04/03/attention.html#background
produces BoolTensor:
[[ True False False]
  [ True  True False]
  [ True  True  True]]

D: The Aladdin code here
https://github.com/aladdinpersson/Machine-Learning-Collection/blob/1d78f390ab8ae803e6a56ef42cad506fc49b2c81/ML/Pytorch/more_advanced/transformer_from_scratch/transformer_from_scratch.py#L56-L57
produces FloatTensor:
[[1. 0. 0.]
   [1. 1. 0.]
   [1. 1. 1.]]
Which Andrew noted gets converted to this (though I have not verified):
[[1. -1e20 -1e20]
   [1. 1. -1e20]
   [1. 1. 1.]]

Implementations A and B are nearly the same and produce identical results.
As Andrew pointed out, implementation D has approximately the same effect.
However, based on the pytorch transformer documentation, A and C have
opposite interpretations. For a BoolTensor, the True locations will not be
attended.

My question is how can we test to be certain that A is the correct usage
and C is incorrect, or vice versa?

**Comment by andrewpeng02** [comment]
Sorry, I had a slight error on my interpretation of implementation D.

D: In the Aladdin code [here](https://github.com/aladdinpersson/Machine-Learning-Collection/blob/1d78f390ab8ae803e6a56ef42cad506fc49b2c81/ML/Pytorch/more_advanced/transformer_from_scratch/transformer_from_scratch.py#L56-L57),
`energy = energy.masked_fill(mask == 0, float("-1e20"))`
Meaning that it changes the weights to -1e20 where mask == 0 and leaves the rest unchanged (remember, lower diagonal are 1's and upper diagonal are 0's), which lines up with implementations A and B.

C:
Here is how the mask is used in Harvard NLP's attention mechanism after being modified in its Batch class to account for padding:
`scores = scores.masked_fill(mask == 0, -1e9)`
This is almost the same as implementation D. It's not too clear to me, but someone in between the tensor probably changed to a numeric tensor, meaning that the Trues (lower diagonal) are converted to 1's and the Falses (upper diagonal) are converted to 0's.

Although the masks are different in their definitions (A and B, C and D alike), the masking of the appropriate values in all four implementations should ultimately result in the same weights. Pytorch adds the mask, while C and D only changes the values where mask == 0. So, all of these implementations are correct even though the mask is defined differently (what to attend and what not to attend are reversed).  

I might look into updating the project requirements (Pytorch -> 1.7) though.

**Comment by dsantiago** [comment]
Sure guys i see the point, maybe my doubt was good to all this brainstorm here... But i will double check next time... Thanks for the clarifications!
